Add POWER8 ASM for 4-way SHA256

TheBlueMatt opened the pull request:
Based on #13191, This speeds up 4-way SHA256 by about 3.75x over the C impl.

laanwj commented:
Cool!
Ping @luke-jr

luke-jr commented on configure.ac:
```diff
@@ -762,6 +762,29 @@ AC_LINK_IFELSE([AC_LANG_SOURCE([
 LDFLAGS="$TEMP_LDFLAGS"
 
+TEMP_LDFLAGS="$LDFLAGS"
+LDFLAGS="$TEMP_LDFLAGS -mpower8-vector"
```
Shouldn't this be CPPFLAGS or CXXFLAGS?

theuni replied:
Agree, but this should be assigned to POWER_VSX_CXXFLAGS or so above, in an AX_CHECK_LINK_FLAG test.

luke-jr commented on src/Makefile.am:
```diff
@@ -309,6 +310,16 @@ crypto_libbitcoin_crypto_avx2_a_CPPFLAGS += -DENABLE_AVX2
 crypto_libbitcoin_crypto_avx2_a_SOURCES = crypto/sha256_avx2.cpp
 
+if HAVE_POWER8
+crypto_libbitcoin_crypto_power8_a_CPPFLAGS = $(AM_CPPFLAGS) -mpower8-vector
```
This is probably needed on CXXFLAGS, since the man page says it changes the code generated.

luke-jr commented on src/crypto/sha256.cpp:
```diff
@@ -29,6 +29,15 @@ namespace sha256d64_avx2
 }
 
+#if defined(__linux__) && defined(HAVE_POWER8)
+#include <sys/auxv.h>
```
I don't think we need this here, and POWER vectoring stuff seems to screw with standard C++ stuff, so maybe not a good idea to have it.

TheBlueMatt replied:
auxv isnt POWER- or vector-specific, its to get getauxval which is a libc function.

luke-jr commented on src/crypto/sha256.cpp:
```diff
@@ -511,6 +520,12 @@ std::string SHA256AutoDetect()
 #endif
     }
+#elif (defined(__linux__)) && defined(HAVE_POWER8)
+    if (getauxval(AT_HWCAP2) & 0x02000000) {
```
Can we move this check to a function in the power8-specific file?

TheBlueMatt replied:
Its right below the x86-specific support check, and its a short file, it seems like overkill.

luke-jr commented on src/crypto/sha256_power8.cpp:
```diff
@@ -0,0 +1,405 @@
+{
+
+typedef __vector uint32_t uint32x4_p8;
+typedef __vector uint8_t uint8x16_p8;
```
`__vector` is a non-standard GCC extension; any reason we can't use `vector` here directly?

luke-jr commented on src/crypto/sha256_power8.cpp:
```diff
@@ -0,0 +1,405 @@
+#include <config/bitcoin-config.h>
+#endif
+
+#ifdef HAVE_POWER8
```
Since this file is only compiled with `HAVE_POWER8`, I think we don't need this here.

luke-jr reviewed (changes requested):
I'd be surprised if this works as-is; either way, some cleanup needed to avoid breaking non-POWER builds.

theuni commented on src/Makefile.am:
```diff
@@ -289,6 +294,32 @@ if USE_ASM
+crypto_libbitcoin_crypto_avx2_a_CPPFLAGS = $(CPPFLAGS)
+if ENABLE_AVX2
+crypto_libbitcoin_crypto_avx2_a_CXXFLAGS += $(AVX2_CXXFLAGS)
+crypto_libbitcoin_crypto_avx2_a_CPPFLAGS += -DENABLE_AVX2
```
Can't compiler defines be used instead so that non-autoconf builds can compile with optims too?

Specifically for this one, rather than ```ifdef(ENABLE_AVX2)```, use ```ifdef(__AVX2__)```

theuni commented on configure.ac:
```diff
@@ -327,6 +330,45 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
 )
 CXXFLAGS="$TEMP_CXXFLAGS"
 
+AC_DEFINE(ENABLE_AVX2, 1, [Define this symbol to build code that uses AVX2 intrinsics])
```
Only define these in their respective check's success cases.

theuni commented on configure.ac:
```diff
@@ -720,6 +762,29 @@ AC_LINK_IFELSE([AC_LANG_SOURCE([
+  }
+  ]])],
+  [
+    AC_DEFINE(HAVE_POWER8,1,[Define if compiler supports POWER8 instructions.])
```
Please move the defines/conditionals down with the other asm flags for consistency.

theuni commented on src/Makefile.am:
```diff
@@ -289,6 +294,32 @@ if USE_ASM
 crypto_libbitcoin_crypto_a_SOURCES += crypto/sha256_sse4.cpp
 endif
 
+crypto_libbitcoin_crypto_sse41_a_CXXFLAGS = $(CXXFLAGS)
```
These should all be $(AM_CXXFLAGS) $(PIE_FLAGS).

theuni commented on src/Makefile.am:
```diff
@@ -289,6 +294,32 @@ if USE_ASM
 endif
 
+crypto_libbitcoin_crypto_sse41_a_CXXFLAGS = $(CXXFLAGS)
+crypto_libbitcoin_crypto_sse41_a_CPPFLAGS = $(CPPFLAGS)
```
Ditto for the CPPFLAGS.

theuni commented on src/Makefile.am:
```diff
@@ -289,6 +294,32 @@ if USE_ASM
+crypto_libbitcoin_crypto_power8_a_SOURCES = crypto/sha256_power8.cpp
+
+LIBBITCOIN_CRYPTO += $(LIBBITCOIN_CRYPTO_POWER8)
+EXTRA_LIBRARIES += $(LIBBITCOIN_CRYPTO_POWER8)
```
append to EXTRA_LIBRARIES outside of the ```if HAVE_POWER8```. These libraries are only built if something depends on them.

theuni commented on src/Makefile.am:
```diff
@@ -289,6 +294,32 @@ if USE_ASM
+crypto_libbitcoin_crypto_power8_a_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
+crypto_libbitcoin_crypto_power8_a_SOURCES = crypto/sha256_power8.cpp
+
+LIBBITCOIN_CRYPTO += $(LIBBITCOIN_CRYPTO_POWER8)
```
I'm not sure whether this works as intended or not, but we don't want to be doing it either way. Add LIBBITCOIN_CRYPTO_POWER8 to bitcoind_LDADD like the others, we can potentially aggregate them as a follow-up.

TheBlueMatt replied:
Why not? Its shorthand for "crypto depends on this" instead of listing it as a dep in every target that may use it. Its also much less britle, which is important given we dont have a POWER travis target.

theuni replied:
I agree with aggregating, but as a clean follow-up, or as part of #13191. I'd just prefer not to have one-off's. It's not clear, for example, if EXTRA_LIBRARIES is appended twice, as it depends on when LIBBITCOIN_CRYPTO is calculated.

theuni reviewed (commented):
Concept ACK!

First set of build-system comments, there are few things to sort out. Will continue reviewing after the first round of cleanups.

TheBlueMatt commented:
@luke-jr No, it definitely works, but its apparently /only/ tested on POWER, not "not yet tested" :p.

theuni commented on configure.ac:
```diff
@@ -327,6 +330,69 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+  ]
+)
+AM_CONDITIONAL([HAVE_POWER8], [$HAVE_POWER8])
+LDFLAGS="$TEMP_LDFLAGS"
```
This should reset CXXFLAGS, not LDFLAGS.

gmaxwell commented:
spiffy! will test soon

TheBlueMatt commented:
Rebased.

TheBlueMatt committed:
Add POWER8 vector impl for 4-way SHA256

This speeds up 4-way SHA256 by about 3.75x over the C impl.

TheBlueMatt committed:
Make checkqueue bench a realistic result

Previously checkqueue bench would essentially just test cross-core
latency with more threads the more threads a host had. Not only did
this make the result worse on hosts with more CPU cores, but it
doesn't mimic how we use the checkqueue in real life, and made
bench_bitcoin almost unusably slow on high-core-count systems,
especially dual-CPU systems.

Instead, we use the same thread count that we use for the real
validation queue, with a 75-microsecond busy-wait on each thread,
which should at least vaguely mimic how we use the checkqueue.

TheBlueMatt commented:
Rebased. Should be ready for merge +/- some more review, though luckily the tests should mostly cover it.

TheBlueMatt commented:
Note that I also had to fix the checkqueue bench as otherwise bench_bitcoin is unsuably slow on multi-cpu (or many-core) systems.

laanwj commented:
I'm unable to test this at the moment. Can someone please post the benchmark results before/after this?

luke-jr commented on src/crypto/sha256_power8.cpp:
```diff
@@ -0,0 +1,401 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+//
+// This is a translation to GCC extended asm syntax from YASM code by Intel
+// (available at the bottom of this file).
```
I don't see any YASM code at the bottom; also, really from Intel?

luke-jr commented on configure.ac:
```diff
@@ -375,6 +376,22 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+    #include <stdint.h>
+  ]], [[
+    unsigned char src[16];
+    __builtin_crypto_vshasigmaw((__vector uint32_t)vec_vsx_ld(0, src), 1, 0xf);
```
Better to use `vector` here. We don't use [the non-standard] `__vector` in the code itself.

luke-jr commented on configure.ac:
```diff
@@ -314,6 +314,7 @@ fi
 AX_CHECK_COMPILE_FLAG([-msse4.2],[[SSE42_CXXFLAGS="-msse4.2"]],,[[$CXXFLAG_WERROR]])
 AX_CHECK_COMPILE_FLAG([-msse4.1],[[SSE41_CXXFLAGS="-msse4.1"]],,[[$CXXFLAG_WERROR]])
 AX_CHECK_COMPILE_FLAG([-mavx -mavx2],[[AVX2_CXXFLAGS="-mavx -mavx2"]],,[[$CXXFLAG_WERROR]])
+AX_CHECK_COMPILE_FLAG([-mpower8-vector],[[POWER8_CXXFLAGS="-mpower8-vector"]],,[[$CXXFLAG_WERROR]])
```
I think `-faltivec` is needed too?

gmaxwell commented:
This needs a trivial rebase but with that works fine against master.  ACK. Will post benchmarks after IBD finishes.

gmaxwell commented:
Before:
MerkleRoot, 5, 800, 21.6763, 0.00541869, 0.00541984, 0.00541886
After:
MerkleRoot, 5, 800, 6.32571, 0.00158138, 0.00158148, 0.00158142

On power9.

Interesting reindex speed was about the same for both.

gmaxwell commented:
@TheBlueMatt  Rebase.

TheBlueMatt commented:
Hmm, well if it doesnt speed up rebase then there is something funky we should look into before merging new code, no?

TheBlueMatt commented:
Gonna go ahead and close it, if someone wants to adopt it and figure out why IBD didn't improve where the equivalent x86 change improved it significantly, they should adopt it.

gmaxwell commented:
The comparable change made a big impact on x86 many months ago-- who knows now... Synchronization primitives are more expensive on power, so it also wouldn't be surprising if our glib use of atomics all over the place weren't moving around the limiting factors in performance on power compared to x86.
